net/http/upstream.go

```diff
@@ -0,0 +1,90 @@
+package http
+
+import "sync/atomic"
+
+type upstream struct {
+	// Current backend index
+	index uint32
+
+	// Balancing parameters
+	currentWeight int32
+	maxWeight     int32
+
+	// preorety backend
+	preoretyBackend *backend
+
+	// List of the upstream backends
+	backends backends
+
+	// greatest common divisor
+	gcd int32
+}
+
+func (ups *upstream) nextBackend(maxRequestsByBackend int) (back *backend) {
+	// First send requests to the priority backend (generally this is the local service)
+	if ups.preoretyBackend != nil {
+		if maxRequestsByBackend <= 0 || maxRequestsByBackend > ups.preoretyBackend.requestCount() {
+			return ups.preoretyBackend
+		}
+	}
+
+	backends := ups.backends
+	backendCount := uint32(len(backends))
+
+	for i := uint32(0); i < backendCount; i++ {
+		index := atomic.AddUint32(&ups.index, 1)
+		back = backends[index%backendCount]
+
+		if maxRequestsByBackend <= 0 || maxRequestsByBackend > back.requestCount() {
+			return back
+		}
+	}
+	return nil
+}
+
+func (ups *upstream) nextWeightBackend(maxRequestsByBackend int) *backend {
+	// First send requests to the priority backend (generally this is the local service)
+	if ups.preoretyBackend != nil {
+		if maxRequestsByBackend <= 0 || maxRequestsByBackend > ups.preoretyBackend.requestCount() {
+			return ups.preoretyBackend
+		}
+	}
+
+	backends := ups.backends
+	backendCount := uint32(len(backends))
+	if backendCount < 1 {
+		return nil
+	}
+
+	for i := uint32(0); i < backendCount; i++ {
+		var (
+			currentWeight int32
+			index         = atomic.AddUint32(&ups.index, 1) % backendCount
+			backend       = backends[index]
+		)
+
+		if index == 0 {
+			if ups.maxWeight != 0 {
+				currentWeight = atomic.AddInt32(&ups.currentWeight, -ups.gcd)
+				if currentWeight <= 0 {
+					atomic.StoreInt32(&ups.currentWeight, ups.maxWeight)
```

Comment:
I think that requests should be distributed among backends proportional to their weights.

The simpliest way to make that is just add a backend to `backends` that much times as required to guarantee the expected distribution (in other words to add a backend to `backends` `weight/gcd` times). So there will not be required to do any loop to iterate over backends, but it will work only with natural weights.

Another way is to use `weight / (float64(totalCount) / float64(backend.requestCount()))` as the weight while pulling a backend. To avoid a loop there could be used a heap structure... but it is too complicated, IMHO

Reply:
I think it must be the separate PR